components/style/logical_geometry.rs

```diff
@@ -570,6 +570,22 @@ impl<T: Add<T, Output = T>> Add for LogicalSize<T> {
     }
 }
 
+// TODO(servo#30577) remove this once underlying bugs are fixed
+impl<T: Add<T, Output = T>> LogicalSize<T> {
+    #[inline]
+    pub fn add_or_warn(self, other: LogicalSize<T>) -> LogicalSize<T> {
+        #[cfg(debug_assertions)]
+        if !(self.debug_writing_mode.mode == other.debug_writing_mode.mode) {
+            log::warn!("debug assertion failed! self.debug_writing_mode.mode == other.debug_writing_mode.mode");
+        }
+        LogicalSize {
+            debug_writing_mode: self.debug_writing_mode,
+            inline: self.inline + other.inline,
+            block: self.block + other.block,
+        }
+    }
+}
+
```

Comment:
Do you mind moving this out of `style` and into the `layout` crate? It's not impossible that we'll remove legacy layout before fixing this one (since legacy layout is a low priority) and we eventually want to push our changes to `style` upstream.

Reply:
hmm, we can’t avoid crashing without some kind of modification to style, because debug_writing_mode is a private field. is there a way we can mark this as not to be upstreamed?

Reply:
Oh, in that case, I guess it is fine to put into `style` for now.